### PR TITLE
Refactor providers package

### PR DESCRIPTION
**Pull request by sebsoto**
This PR contains commits relating to refactoring the providers packages, used by the WMCO e2e tests to generate MachineSet specs. This will make adding additional platform easier.

### CODE REVIEW

File: test/e2e/providers/base/base.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	mapi "github.com/openshift/api/machine/v1beta1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/windows-machine-config-operator/test/e2e/clusterinfo"
+)
+
+// New returns a new MachineSet base, which can be customized for each individual provider
+func New(infrastructureName string, replicas int32, withWindowsLabel bool) *mapi.MachineSet {
```
Comment (saifshaikh48) [inline test/e2e/providers/base/base.go:11]: Consider renaming to `NewMachineSet`

File: test/e2e/providers/azure/azure.go [hunk 2]
```diff
@@ -109,62 +110,8 @@ func (p *Provider) GenerateMachineSet(withWindowsLabel bool, replicas int32) (*m
...
-						Value: &runtime.RawExtension{
-							Raw: rawProviderSpec,
-						},
-					},
-				},
-			},
-		},
-	}
+	machineSet := base.New(p.status.InfrastructureName, replicas, withWindowsLabel)
+	machineSet.Spec.Template.Spec.ProviderSpec.Value = &runtime.RawExtension{Raw: rawProviderSpec}
```
Comment (saifshaikh48) [inline test/e2e/providers/azure/azure.go:114]: Why not do this within the function that creates the MachineSet?

File: test/e2e/providers/aws/aws.go [hunk 3]
```diff
@@ -32,6 +32,8 @@ const (
 )
 
 type awsProvider struct {
+	// status contains information about the cluster infrastructure
+	status *config.InfrastructureStatus
```
Comment (aravindhp) [inline test/e2e/providers/aws/aws.go:35]: I suggest just embedding this.

File: test/e2e/providers/azure/azure.go [hunk 4]
```diff
@@ -31,38 +31,21 @@ const (
...
-	if status.Azure == nil || status.Azure == (&config.AzurePlatformStatus{}) {
-		return nil, fmt.Errorf("azure platform status is nil")
-	}
-	if status.Azure.NetworkResourceGroupName == "" {
-		return nil, fmt.Errorf("azure network resource group name is empty")
-	}
-	rg := status.Azure.ResourceGroupName
-	netrg := status.Azure.NetworkResourceGroupName
-
+func (p *Provider) newAzureMachineProviderSpec(location, zone string) (*mapi.AzureMachineProviderSpec, error) {
```
Comment (aravindhp) [inline test/e2e/providers/azure/azure.go:48]: Now that you are reworking this, I suggest naming this `getAzureMachineProviderSpect`

File: test/e2e/providers/shared/shared.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
+package shared
```
Comment (aravindhp) [inline test/e2e/providers/shared/shared.go:1]: Rather than `shared` I suggest calling this `machineset` given that is what it is dealing with.

**Review by aravindhp — CHANGES_REQUESTED** [review]
Nice work, @sebsoto. Just a few naming suggestions.

File: test/e2e/providers/azure/azure.go (hunk 4 above)
Comment (sebsoto) [reply test/e2e/providers/azure/azure.go:48]: The function is actually constructing a new AzureMachineProviderSpec based on the input, to me it feels the `new` prefix fits better

Comment (aravindhp) [reply test/e2e/providers/azure/azure.go:48]: I see your point. Leave it as is then.

File: test/e2e/providers/aws/aws.go [hunk 6]
```diff
@@ -61,9 +61,9 @@ func newSession(credentialPath, credentialAccountID, region string) (*awssession
 // credentialPath is the file path the AWS credentials file.
 // credentialAccountID is the account name the user uses to create VM instance.
 // The credentialAccountID should exist in the AWS credentials file pointing at one specific credential.
-func newAWSProvider(openShiftClient *clusterinfo.OpenShift, credentialPath,
-	credentialAccountID, instanceType, region string) (*awsProvider, error) {
-	session, err := newSession(credentialPath, credentialAccountID, region)
+func newAWSProvider(openShiftClient *clusterinfo.OpenShift, status *config.InfrastructureStatus, credentialPath,
```
Comment (aravindhp) [inline test/e2e/providers/aws/aws.go:63]: Sorry forgot to call this out before. I suggest making this infraStatus to be explicit. `status` is too generic.

**Review by aravindhp — APPROVED** [review]

File: test/e2e/providers/aws/aws.go [hunk 7]
```diff
@@ -32,6 +32,8 @@ const (
 )
 
 type awsProvider struct {
+	// status contains information about the cluster infrastructure
```
Comment (saifshaikh48) [inline test/e2e/providers/aws/aws.go:34]: nit drop `status` from this doc

File: test/e2e/providers/aws/aws.go [hunk 8]
```diff
@@ -76,44 +76,28 @@ func newAWSProvider(openShiftClient *clusterinfo.OpenShift, credentialPath,
...
 
-// SetupAWSCloudProvider creates AWS provider using the give OpenShift client
-// This is the first step of the e2e test and fails the test upon error.
-func SetupAWSCloudProvider(region string) (*awsProvider, error) {
-	oc, err := clusterinfo.GetOpenShift()
-	if err != nil {
-		return nil, fmt.Errorf("failed to initialize OpenShift client with error: %v", err)
-	}
+// New returns a new awsProvider
+func New(oc *clusterinfo.OpenShift, status *config.InfrastructureStatus) (*awsProvider, error) {
```
Comment (saifshaikh48) [inline test/e2e/providers/aws/aws.go:86]: `status` param --> `infraStatus` ?

**Pull request merged**